### PR TITLE
Map Testlib To Test with owner tag heuristic

### PR DESCRIPTION
**Pull request by sfc-gh-gguo**
This change is to map testlib java_library to actual java_test with tag heuristic. The test target cannot be found from reverse dependency chain because an item in the chain is invisible to IDE to avoid excessive indexing.

Adding log here shouldn't bring too much overhead as this is triggered only when user open a test file.

### CODE REVIEW

**Comment by jmmv** [comment]
Are you trying to find the test target for a specific source file?

You should talk to Jason. He was trying to do something similar for ownership attribution, and he added some tags to help with this that we'd presumably query here.

**Comment by sfc-gh-gguo** [comment]
> Are you trying to find the test target for a specific source file?
> 
> You should talk to Jason. He was trying to do something similar for ownership attribution, and he added some tags to help with this that we'd presumably query here.

Updated the logic to use the owner tag Kanglei added for the java-deps; didn't originally go this approach because I wasn't sure if the tags are preserved in memory for Google plugin, and it turns out they are. 

Question comes to this mystery again: why is the Jetbrains plugin stripping information, but still storing so much larger target information file that causes freeze? That keeps coming onto the top of my mind.

File: base/src/com/google/idea/blaze/base/run/testmap/FilteredTargetMap.java [hunk 1]
```diff
@@ -90,6 +95,40 @@ private ImmutableSet<TargetIdeInfo> targetsForSourceFilesImpl(
       if (filter.test(target)) {
         result.add(target);
       }
+
+      // Tag-based heuristic: if this target has an owner: tag, map to the owner test target
+      // Tags format: "owner://path/to:actual-test-target"
+      if (target != null) {
+        target.getTags().stream()
```
Comment (sfc-gh-kafang) [inline base/src/com/google/idea/blaze/base/run/testmap/FilteredTargetMap.java]: just a bit sad that this depends on the tagging in monorepo and has to work together. But at least the tag on this testlib macro has been there for a long time.

Remind me again, the original issue is that test -> combined_deps -> test_lib (generated by macro) -> java test file, since `combined_deps` is `no-ide`, the association gets lost in the middle.

This operate on test_lib and directly gets the test association via the tagging on test_lib, which is fine.
 
Does no_src test work for this too? the test_lib simply won't get generated, so it's:
test (no src) -> combined_deps (no-ide) -> testlib (BFG gen'd library) -> java test file
Since tags are not generated on BFG gen'd library, wouldn't it still be blocked?

Comment (sfc-gh-gguo) [reply base/src/com/google/idea/blaze/base/run/testmap/FilteredTargetMap.java]: No it doesn't work for those. Testlibs there are separately created as -test-lib. Might have to add naming heuristic to support them, or re-audit all such cases and add the owner tag. 

I suggest doing naming heuristic for now to unblock developers.

File: base/src/com/google/idea/blaze/base/run/testmap/FilteredTargetMap.java [hunk 2]
```diff
@@ -90,6 +96,27 @@ private ImmutableSet<TargetIdeInfo> targetsForSourceFilesImpl(
       if (filter.test(target)) {
         result.add(target);
       }
+
+      // Naming heuristic: if this is a testlib target (e.g., .testlib or -test-lib),
+      // also try to find and add the corresponding test target without the suffix
```
Comment (sfc-gh-kafang) [inline base/src/com/google/idea/blaze/base/run/testmap/FilteredTargetMap.java]:
```
// Naming heuristic: Today we have two different kind of intermediate test libraries in snowflake codebase:
// - `-test-lib` generated by BFG
// - `.testlib` generated by `java_monolith_it_test` macro 
// These intermediate library may lose association with their test binary target due to macro implementation,
// Here we use heuristic to locate the java_test binary rule of interest.
// Also try to find and add the corresponding test target without the suffix
```

File: base/src/com/google/idea/blaze/base/run/testmap/FilteredTargetMap.java [hunk 3]
```diff
@@ -90,6 +96,27 @@ private ImmutableSet<TargetIdeInfo> targetsForSourceFilesImpl(
...
+      String targetName = targetKey.getLabel().targetName().toString();
```
Comment (sfc-gh-kafang) [inline base/src/com/google/idea/blaze/base/run/testmap/FilteredTargetMap.java:106]: This is a bit different than the previous owner finding logic, we would unconditionally trying to associate the target here? How about the ones that are not impacted by the combine_deps? would they execute this as well?

Is the current logic simply collecting a bunch of possibilities and deal with them later?

Comment (sfc-gh-gguo) [reply base/src/com/google/idea/blaze/base/run/testmap/FilteredTargetMap.java:106]: It's actually the same, just this one is written in loop and I haven't request Copilot to change this to stream yet at this commit. It's adding the owner target to `todo.add(baseTargetKey)`, and then the existing logic will walk it through its filter  `if (filter.test(target)) {` with a visited tracking `if (!seen.add(targetKey)) {`.

Comment (sfc-gh-gguo) [reply base/src/com/google/idea/blaze/base/run/testmap/FilteredTargetMap.java:106]: This is the filter btw:
 ```
       target ->
            target.isPlainTarget()
                && target.getKind().hasLanguage(LanguageClass.JAVA)
                && target.getKind().getRuleType().equals(RuleType.BINARY));
```
https://github.com/Snowflake-Labs/intellij/blob/013bda581e6242aed2648b4e5919bd0275a74779/java/src/com/google/idea/blaze/java/run/producers/JavaBinaryContextProvider.java#L230-L233

**Review by sfc-gh-kafang — APPROVED** [review]
Accepting assuming the following cases pass:
- java_it_test with src
- java_it_test without src
- java_junit5_test with src
- java_junit5_test wihtout src

If you can think other edge cases do sanity test on those too
